src/lightning/data/datasets/iterable.py

```diff
@@ -0,0 +1,447 @@
+        (e.g. loading the previously downloaded file into memory)
+    - ``load_sample_from_chunk`` implements how to retrieve a single sample from the current chunk
+        (e.g. indexing the chunk if it's a list or just returning it if the chunk has a size of 1)
+
+    Args:
+        chunks: The chunked_data to load.
+        chunk_size: The number of samples in each chunk
+        num_parallel_chunks: How many chunks to load in parallel.
+        env: The compute-environment. Important for sharding. Contains the distributed world-size,
+            the distributed global rank, the number of workers on the current rank and the current worker rank.
+            If None, it will try to detect these things automatically.
+        shuffle: Whether to shuffle your data. Will shuffle both, order of chunks before sharding and order of
+            samples within each chunk.
+        seed: The seed for the random number generator. If :param:`shuffle` = False, the seed has no effect.
+        wrap: Whether to restart your dataset if it's exhausted. If set to True, it results in a
+            virtually infinite dataset looping through the same data over and over again.
+        lazy_shuffle: Whether to shuffle your data lazily instead of upfront.
+            This consumes a lot less memory, but may yield undeterministic results.
+
+        Note:
+            :param:`lazy_shuffle` is experimental, consumes less memory than shuffling everything in advance (default)
+            but may result in undeterministic behavior.
+
+        Note:
+            On resume from a state-dict, we always skip currently started chunks as these would make the data-order
+            impossible to determine with sharding. Upon resuming from a point, where a new chunk would be started
+            anyways, nothing is skipped.
+
+        Note:
+            Order of data is only guaranteed when resuming with the same distributed settings and the same number of
+            workers. Everything else leads to different sharding and therefore results in different data order.
+    """
+
+    # TODO: backend docstring
+    def __init__(
+        self,
+        chunks: Sequence[Any],
+        chunk_size: int = 1,
+        num_parallel_chunks: int = 1,
+        env: Optional[Environment] = None,
+        shuffle: bool = False,
+        seed: Optional[int] = None,
+        wrap: bool = False,
+        lazy_shuffle: bool = False,
+        backend: str = "local",
+    ):
+        _StatefulIterableDataset.__init__(self)
+        _Dataset.__init__(self, backend=backend)
+
+        chunks = [_Chunk(c, chunk_size=chunk_size) for c in chunks]
+        if env is None:
+            # must detect distributed env here since distributed is not initialized in worker processes with ddp spawn
+            # can't detect worker env here since workers not yet initialized
+            env = Environment(DistributedEnv.detect(), None)
```

Comment:
I understand this only partially. Why can we assume the dist env is initialized here? The creation of the dataset might happen in the main process before processes have been spawned and dist is init. I suggest to delay the creation of env entirely to `__iter__`. It looks like `self._env` is not needed during the dataset init.

Reply:
I had a problem initializing it in `__iter__` since the torch.distributed isn't initialized in the worker processes.